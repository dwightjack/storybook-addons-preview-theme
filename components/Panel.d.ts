import * as React from 'react';
export interface PanelProp {
    api: any;
    active: boolean;
    channel: any;
}
export declare function Panel({ api, active, channel, }: {
    api: any;
    active: any;
    channel: any;
}): React.FunctionComponentElement<PanelProp>;
