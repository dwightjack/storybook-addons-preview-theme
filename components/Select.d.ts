import * as React from 'react';
import { Theme } from '../types';
export interface SelectProps {
    onChange: () => void;
    options: Theme[];
    value?: string;
}
export declare function Select({ onChange, options, value, }: {
    onChange: any;
    options: any;
    value?: string;
}): React.FunctionComponentElement<SelectProps>;
