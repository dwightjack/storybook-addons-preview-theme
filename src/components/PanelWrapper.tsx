import { styled } from '@storybook/theming';

export const PanelWrapper = styled.div`
  height: 100%;
  overflow: auto;
  width: 100%;
`;
