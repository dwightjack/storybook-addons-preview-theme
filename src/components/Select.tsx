import * as React from 'react';
import { Form } from '@storybook/components';
import { Theme } from '../types';

export interface SelectProps {
  onChange: () => void;
  options: Theme[];
  value?: string;
}

export function Select({
  onChange,
  options,
  value = '',
}): React.FunctionComponentElement<SelectProps> {
  return (
    <Form.Select
      onChange={onChange}
      value={value}
      name="preview-theme"
      label="Preview Theme"
      size="flex"
    >
      {[
        { name: 'empty', className: 'none', label: 'select...' },
        ...options,
      ].map(({ name, label, className }) => (
        <option value={className} key={name}>
          {label}
        </option>
      ))}
    </Form.Select>
  );
}
