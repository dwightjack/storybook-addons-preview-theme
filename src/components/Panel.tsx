import * as React from 'react';
import { useState, useEffect } from 'react';
import { Form, Placeholder } from '@storybook/components';
import { STORY_CHANGED } from '@storybook/core-events';
import { Select } from './Select';
import { PanelWrapper } from './PanelWrapper';
import { getThemes } from '../utils';

export interface PanelProp {
  api: any;
  active: boolean;
  channel: any;
}

export function Panel({
  api,
  active,
  channel,
}): React.FunctionComponentElement<PanelProp> {
  const [selected, setSelected] = useState('none');
  const data = api.getCurrentStoryData();

  useEffect(() => {
    const listener = (): void => {
      setSelected('none');
    };

    api.on(STORY_CHANGED, listener);

    return () => {
      api.off(STORY_CHANGED, listener);
    };
  }, []);

  useEffect(() => {
    channel.emit('preview-theme:change', selected);
  }, [selected]);

  if (!active || !data) {
    return null;
  }

  const themes = getThemes(data.parameters || {});

  if (!Array.isArray(themes) || themes.length === 0) {
    return (
      <Placeholder>Theme swatches not available for this story</Placeholder>
    );
  }

  return (
    <PanelWrapper>
      <Form>
        <Form.Field>
          <Select
            value={selected}
            onChange={({ target }) => setSelected(target.value)}
            options={themes}
          />
        </Form.Field>
      </Form>
    </PanelWrapper>
  );
}
