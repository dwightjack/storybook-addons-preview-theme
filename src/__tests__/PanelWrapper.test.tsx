import * as React from 'react';
import { render } from 'react-testing-library';
import { PanelWrapper } from '../components/PanelWrapper';

describe('<PanelWrapper />', () => {
  test('matches snapshot', () => {
    const { container } = render(<PanelWrapper />);
    expect(container.firstElementChild).toMatchSnapshot();
  });
});
