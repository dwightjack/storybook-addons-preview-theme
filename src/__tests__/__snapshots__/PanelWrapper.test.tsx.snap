// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PanelWrapper /> matches snapshot 1`] = `
<div
  class="css-13bbivi"
/>
`;
